File: 4-median-of-two-sorted-arrays/median-of-two-sorted-arrays.rs
impl Solution {
    pub fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
        // Ensure nums1 is the shorter array to optimize search
        if nums1.len() > nums2.len() {
            return Self::find_median_sorted_arrays(nums2, nums1);
        }
        
        let x = nums1.len();
        let y = nums2.len();
        let total_length = x + y;
        
        // Binary search boundaries
        let mut low = 0;
        let mut high = x;
        
        while low <= high {
            // Partition positions for both arrays
            let partition_x = (low + high) / 2;
            let partition_y = (total_length + 1) / 2 - partition_x;
            
            // Get the four elements around the partition
            let max_x_left = if partition_x == 0 { 
                std::i32::MIN 
            } else { 
                nums1[partition_x - 1] 
            };
            
            let min_x_right = if partition_x == x { 
                std::i32::MAX 
            } else { 
                nums1[partition_x] 
            };
            
            let max_y_left = if partition_y == 0 { 
                std::i32::MIN 
            } else { 
                nums2[partition_y - 1] 
            };
            
            let min_y_right = if partition_y == y { 
                std::i32::MAX 
            } else { 
                nums2[partition_y] 
            };
            
            // Found the correct partition
            if max_x_left <= min_y_right && max_y_left <= min_x_right {
                // Even total length
                if total_length % 2 == 0 {
                    return (max_x_left.max(max_y_left) as f64 + min_x_right.min(min_y_right) as f64) / 2.0;
                } 
                // Odd total length
                else {
                    return max_x_left.max(max_y_left) as f64;
                }
            } 
            // Adjust the binary search
            else if max_x_left > min_y_right {
                high = partition_x - 1;
            } else {
                low = partition_x + 1;
            }
        }
        
        // If we get here, the input arrays are not sorted or contain invalid elements
        0.0 // Default error case (shouldn't reach here with valid inputs)
    }
}

File: 4-median-of-two-sorted-arrays/median-of-two-sorted-arrays.php
class Solution {

    /**
     * @param Integer[] $nums1
     * @param Integer[] $nums2
     * @return Float
     */
    function findMedianSortedArrays($nums1, $nums2) {
        // Ensure nums1 is the smaller array to perform binary search on it
        if (count($nums1) > count($nums2)) {
            return $this->findMedianSortedArrays($nums2, $nums1);
        }
        
        $m = count($nums1);
        $n = count($nums2);
        $left = 0;
        $right = $m;
        $totalLeft = ($m + $n + 1) / 2;  // Total elements in left partition
        
        while ($left <= $right) {
            $i = intval(($left + $right) / 2);  // Partition position in nums1
            $j = intval($totalLeft - $i);       // Partition position in nums2
            
            // Handle edge cases where i or j is out of bounds
            $nums1LeftMax = ($i == 0) ? PHP_INT_MIN : $nums1[$i - 1];
            $nums1RightMin = ($i == $m) ? PHP_INT_MAX : $nums1[$i];
            $nums2LeftMax = ($j == 0) ? PHP_INT_MIN : $nums2[$j - 1];
            $nums2RightMin = ($j == $n) ? PHP_INT_MAX : $nums2[$j];
            
            if ($nums1LeftMax <= $nums2RightMin && $nums2LeftMax <= $nums1RightMin) {
                // Found the correct partition
                if (($m + $n) % 2 == 1) {
                    return max($nums1LeftMax, $nums2LeftMax);
                } else {
                    return (max($nums1LeftMax, $nums2LeftMax) + min($nums1RightMin, $nums2RightMin)) / 2;
                }
            } elseif ($nums1LeftMax > $nums2RightMin) {
                $right = $i - 1;
            } else {
                $left = $i + 1;
            }
        }
        
        return 0.0;  // Shouldn't reach here for valid inputs
    }
}